<a href="https://colab.research.google.com/github/qjpbtuzara-ai/CPE311-CPE22S3/blob/main/Another_copy_of_Hands_on_Activity_1_1_Optimization_and_Knapsack_Problem_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Hands-on Activity 1.1 | Optimization and Knapsack Problem

#### Objective(s):

This activity aims to demonstrate how to apply  greedy and brute force algorithms to solve optimization problems

#### Intended Learning Outcomes (ILOs):
* Demonstrate how to solve knapsacks problems using greedy algorithm
* Demonstrate how to  solve knapsacks problems using brute force algorithm


#### Resources:
* Jupyter Notebook


#### Procedures:

1. Create a Food class that defines the following:
* name of the food
* value of the food
* calories of the food

2. Create the following methods inside the Food class:
* A method that returns the value of the food
* A method that returns the cost of the food
* A method that calculates the density of the food (Value / Cost)
* A method that returns a string to display the name, value and calories of the food

In [2]:
class Food(object):
    def __init__(self, n, v, w):
        # Make the variables private
        self.name = n
        self.value = v
        self.calories = w
    def getValue(self):
        return self.value
    def getCost(self):
        return self.calories
    def density(self):
        return self.getValue()/self.getCost()
    def __str__(self):
        return self.name + ': <' + str(self.value)+ ', ' + str(self.calories) + '>'

3. Create a buildMenu method that builds the name, value and calories of the food


In [3]:
def buildMenu(names, values, calories):
    menu = []
    for i in range(len(values)):
        menu.append(Food(names[i], values[i],calories[i]))
    return menu

4. Create a method greedy to return total value and cost of added food based on the desired maximum cost

In [4]:
def greedy(items, maxCost, keyFunction):
    """Assumes items a list, maxCost >= 0,         keyFunction maps elements of items to numbers"""
    itemsCopy = sorted(items, key = keyFunction,
                       reverse = True)
    result = []
    totalValue, totalCost = 0.0, 0.0
    for i in range(len(itemsCopy)):
        if (totalCost+itemsCopy[i].getCost()) <= maxCost:
            result.append(itemsCopy[i])
            totalCost += itemsCopy[i].getCost()
            totalValue += itemsCopy[i].getValue()
    return (result, totalValue)

5. Create a testGreedy method to test the greedy method

In [5]:
def testGreedy(items, constraint, keyFunction):
    taken, val = greedy(items, constraint, keyFunction)
    print('Total value of items taken =', val)
    for item in taken:
        print('   ', item)

In [6]:
def testGreedys(foods, maxUnits):
    print('Use greedy by value to allocate', maxUnits,          'calories')
    testGreedy(foods, maxUnits, Food.getValue)
    print('\nUse greedy by cost to allocate', maxUnits,          'calories')
    testGreedy(foods, maxUnits, lambda x: 1/Food.getCost(x))
    print('\nUse greedy by density to allocate', maxUnits,          'calories')
    testGreedy(foods, maxUnits, Food.density)

6. Create arrays of food name, values and calories
7. Call the buildMenu to create menu for food
8. Use testGreedys method to pick food according to the desired calories

In [7]:
names = ['wine', 'beer', 'pizza', 'burger', 'fries','cola', 'apple', 'donut', 'cake']
values = [89,90,95,100,90,79,50,10]
calories = [123,154,258,354,365,150,95,195]
foods = buildMenu(names, values, calories)
testGreedys(foods, 2000)

Use greedy by value to allocate 2000 calories
Total value of items taken = 603.0
    burger: <100, 354>
    pizza: <95, 258>
    beer: <90, 154>
    fries: <90, 365>
    wine: <89, 123>
    cola: <79, 150>
    apple: <50, 95>
    donut: <10, 195>

Use greedy by cost to allocate 2000 calories
Total value of items taken = 603.0
    apple: <50, 95>
    wine: <89, 123>
    cola: <79, 150>
    beer: <90, 154>
    donut: <10, 195>
    pizza: <95, 258>
    burger: <100, 354>
    fries: <90, 365>

Use greedy by density to allocate 2000 calories
Total value of items taken = 603.0
    wine: <89, 123>
    beer: <90, 154>
    cola: <79, 150>
    apple: <50, 95>
    pizza: <95, 258>
    burger: <100, 354>
    fries: <90, 365>
    donut: <10, 195>


Task 1: Change the maxUnits to 100

In [14]:
names = ['wine', 'beer', 'pizza', 'burger', 'fries','cola', 'apple', 'donut', 'cake']
values = [89,90,95,100,90,79,50,10]
calories = [123,154,258,354,365,150,95,195]
foods = buildMenu(names, values, calories)
testGreedys(foods, 100)

Use greedy by value to allocate 100 calories
Total value of items taken = 50.0
    apple: <50, 95>

Use greedy by cost to allocate 100 calories
Total value of items taken = 50.0
    apple: <50, 95>

Use greedy by density to allocate 100 calories
Total value of items taken = 50.0
    apple: <50, 95>


Task 2: Modify codes to add additional weight (criterion) to select food items.

In [15]:
class Food(object):
    def __init__(self, n, v, w, f):
        self.name = n
        self.value = v
        self.calories = w
        self.fat = f

    def getValue(self):
        return self.value

    def getCost(self):
        return self.calories

    def getFat(self):
        return self.fat

    def density(self):
        if self.calories == 0:
            return float('inf')
        return self.getValue() / self.getCost()

    def fat_density(self):
        if self.fat == 0:
            return float('inf')
        return self.getValue() / self.getFat()

    def __str__(self):

        return self.name + ': <' + str(self.value) + ', ' + str(self.calories) + ', ' + str(self.fat) + '>'

def buildMenu(names, values, calories, fats):
    menu = []
    for i in range(len(values)):
        menu.append(Food(names[i], values[i], calories[i], fats[i]))
    return menu


names_updated = ['wine', 'beer', 'pizza', 'burger', 'fries','cola', 'apple', 'donut', 'cake']
values_updated = [89,90,95,100,90,79,50,10, 70]
calories_updated = [123,154,258,354,365,150,95,195, 300]
fats_new = [10.5, 15.2, 30.1, 45.8, 40.7, 8.9, 5.3, 25.0, 35.5]

foods_new_criterion = buildMenu(names_updated, values_updated, calories_updated, fats_new)

def testGreedys_with_additional_criterion(foods, maxUnits):
    print('--- Testing with additional criterion (Fat) ---')

    print('\nUse greedy by value to allocate', maxUnits, 'calories')
    testGreedy(foods, maxUnits, Food.getValue)

    print('\nUse greedy by cost (calories) to allocate', maxUnits, 'calories')
    testGreedy(foods, maxUnits, lambda x: 1/Food.getCost(x))

    print('\nUse greedy by density (value/calories) to allocate', maxUnits, 'calories')
    testGreedy(foods, maxUnits, Food.density)

    print('\nUse greedy by fat density (value/fat) to allocate', maxUnits, 'calories')
    testGreedy(foods, maxUnits, Food.fat_density)

    print('\nUse greedy by inverse fat (least fat first) to allocate', maxUnits, 'calories')
    testGreedy(foods, maxUnits, lambda x: float('inf') if x.getFat() == 0 else 1/x.getFat())


testGreedys_with_additional_criterion(foods_new_criterion, 1000)

--- Testing with additional criterion (Fat) ---

Use greedy by value to allocate 1000 calories
Total value of items taken = 424.0
    burger: <100, 354, 45.8>
    pizza: <95, 258, 30.1>
    beer: <90, 154, 15.2>
    wine: <89, 123, 10.5>
    apple: <50, 95, 5.3>

Use greedy by cost (calories) to allocate 1000 calories
Total value of items taken = 413.0
    apple: <50, 95, 5.3>
    wine: <89, 123, 10.5>
    cola: <79, 150, 8.9>
    beer: <90, 154, 15.2>
    donut: <10, 195, 25.0>
    pizza: <95, 258, 30.1>

Use greedy by density (value/calories) to allocate 1000 calories
Total value of items taken = 413.0
    wine: <89, 123, 10.5>
    beer: <90, 154, 15.2>
    cola: <79, 150, 8.9>
    apple: <50, 95, 5.3>
    pizza: <95, 258, 30.1>
    donut: <10, 195, 25.0>

Use greedy by fat density (value/fat) to allocate 1000 calories
Total value of items taken = 413.0
    apple: <50, 95, 5.3>
    cola: <79, 150, 8.9>
    wine: <89, 123, 10.5>
    beer: <90, 154, 15.2>
    pizza: <95, 258, 30.1>
   

Task 3: Test your modified code to test the greedy algorithm to select food items with your additional weight.

In [16]:
names_updated = ['wine', 'beer', 'pizza', 'burger', 'fries','cola', 'apple', 'donut', 'cake']
values_updated = [89,90,95,100,90,79,50,10, 70] # Added a value for 'cake'
calories_updated = [123,154,258,354,365,150,95,195, 300] # Added calories for 'cake'
fats_new = [10.5, 15.2, 30.1, 45.8, 40.7, 8.9, 5.3, 25.0, 35.5] # Example fat content for each item

foods_new_criterion = buildMenu(names_updated, values_updated, calories_updated, fats_new)
testGreedys_with_additional_criterion(foods_new_criterion, 1000)

--- Testing with additional criterion (Fat) ---

Use greedy by value to allocate 1000 calories
Total value of items taken = 424.0
    burger: <100, 354, 45.8>
    pizza: <95, 258, 30.1>
    beer: <90, 154, 15.2>
    wine: <89, 123, 10.5>
    apple: <50, 95, 5.3>

Use greedy by cost (calories) to allocate 1000 calories
Total value of items taken = 413.0
    apple: <50, 95, 5.3>
    wine: <89, 123, 10.5>
    cola: <79, 150, 8.9>
    beer: <90, 154, 15.2>
    donut: <10, 195, 25.0>
    pizza: <95, 258, 30.1>

Use greedy by density (value/calories) to allocate 1000 calories
Total value of items taken = 413.0
    wine: <89, 123, 10.5>
    beer: <90, 154, 15.2>
    cola: <79, 150, 8.9>
    apple: <50, 95, 5.3>
    pizza: <95, 258, 30.1>
    donut: <10, 195, 25.0>

Use greedy by fat density (value/fat) to allocate 1000 calories
Total value of items taken = 413.0
    apple: <50, 95, 5.3>
    cola: <79, 150, 8.9>
    wine: <89, 123, 10.5>
    beer: <90, 154, 15.2>
    pizza: <95, 258, 30.1>
   

9. Create method to use  Bruteforce algorithm instead of greedy algorithm

In [11]:
def maxVal(toConsider, avail):
    """Assumes toConsider a list of items, avail a weight
       Returns a tuple of the total value of a solution to the
         0/1 knapsack problem and the items of that solution"""
    if toConsider == [] or avail == 0:
        result = (0, ())
    elif toConsider[0].getCost() > avail:
        #Explore right branch only
        result = maxVal(toConsider[1:], avail)
    else:
        nextItem = toConsider[0]
        #Explore left branch
        withVal, withToTake = maxVal(toConsider[1:],
                                     avail - nextItem.getCost())
        withVal += nextItem.getValue()
        #Explore right branch
        withoutVal, withoutToTake = maxVal(toConsider[1:], avail)
        #Choose better branch
        if withVal > withoutVal:
            result = (withVal, withToTake + (nextItem,))
        else:
            result = (withoutVal, withoutToTake)
    return result

In [12]:
def testMaxVal(foods, maxUnits, printItems = True):
    print('Use search tree to allocate', maxUnits,
          'calories')
    val, taken = maxVal(foods, maxUnits)
    print('Total costs of foods taken =', val)
    if printItems:
        for item in taken:
            print('   ', item)

In [13]:
names = ['wine', 'beer', 'pizza', 'burger', 'fries','cola', 'apple', 'donut', 'cake']
values = [89,90,95,100,90,79,50,10]
calories = [123,154,258,354,365,150,95,195]
foods = buildMenu(names, values, calories)
testMaxVal(foods, 2400)

Use search tree to allocate 2400 calories
Total costs of foods taken = 603
    donut: <10, 195>
    apple: <50, 95>
    cola: <79, 150>
    fries: <90, 365>
    burger: <100, 354>
    pizza: <95, 258>
    beer: <90, 154>
    wine: <89, 123>


#### Supplementary Activity:

In [17]:
def solve_knapsack_raw():
    items = [
        {"name": "Medical Supplies", "w": 10, "v": 60},
        {"name": "Food Rations", "w": 20, "v": 100},
        {"name": "Water Purifiers", "w": 30, "v": 120}
    ]
    capacity = 50

    def get_all_combinations(arr):
        if not arr:
            return [[]]
        first = arr[0]
        rest = arr[1:]
        combs_without_first = get_all_combinations(rest)
        combs_with_first = []
        for lobby in combs_without_first:
            combs_with_first.append(lobby + [first])
        return combs_without_first + combs_with_first

    all_combos = get_all_combinations(items)

    bf_best_value = 0
    bf_best_items = []

    for combo in all_combos:
        current_weight = sum(item['w'] for item in combo)
        current_value = sum(item['v'] for item in combo)

        if current_weight <= capacity and current_value > bf_best_value:
            bf_best_value = current_value
            bf_best_items = [item['name'] for item in combo]


    greedy_items = sorted(items, key=lambda x: x['v']/x['w'], reverse=True)

    g_total_value = 0
    g_total_weight = 0
    g_chosen_items = []

    for item in greedy_items:
        if g_total_weight + item['w'] <= capacity:
            g_chosen_items.append(item['name'])
            g_total_weight += item['w']
            g_total_value += item['v']

    return bf_best_items, bf_best_value, g_chosen_items, g_total_value

bf_items, bf_val, g_items, g_val = solve_knapsack_raw()

print("--- BRUTE FORCE RESULT ---")
print(f"Items: {bf_items}")
print(f"Total Value: {bf_val}")

print("\n--- GREEDY RESULT ---")
print(f"Items: {g_items}")
print(f"Total Value: {g_val}")

--- BRUTE FORCE RESULT ---
Items: ['Water Purifiers', 'Food Rations']
Total Value: 220

--- GREEDY RESULT ---
Items: ['Medical Supplies', 'Food Rations']
Total Value: 160


* Choose a real-world problem that solves knapsacks problem
* Use the greedy and brute force algorithm to solve knapsacks problem


#### Conclusion:

#type your conclusion here

I have learned ways in which to handle the knapsack problem and ways to optimize it. The 2 algorithms in question would be the brute force algorithm in which every possible outcome is explored and will be always guaranteed and the greedy algorithm in which it finds a good option which may not be the best but will be faster than the brute force algorithm. The brute force algorithm may be slower but it finds the best solution to be exact and the greedy found 160 because it didnt realize that the best item was actually a poor fit for the total capacity. In short, brute force is for certainty while greedy is for efficiency.